apps/aecore/src/aec_chain_state.erl

```diff
@@ -721,51 +783,91 @@ assert_state_hash_valid(Trees, Node) ->
     case is_micro_block(Node) of
         true ->
+            #fork_info{fees = FeesIn} = ForkInfo,
             apply_micro_block_transactions(Node, FeesIn, Trees);
         false ->
+            #fork_info{fees = FeesIn, fraud = FraudStatus} = ForkInfo,
             GasFees = calculate_gas_fee(aec_trees:calls(Trees)),
             TotalFees = GasFees + FeesIn,
             Trees1 = aec_trees:perform_pre_transformations(Trees, node_height(Node)),
             Delay  = aec_governance:beneficiary_reward_delay(),
             case node_height(Node) > aec_block_genesis:height() + Delay of
-                true  -> {grant_fees(Node, Trees1, Delay, TotalFees, State), TotalFees};
+                true  -> {grant_fees(Node, Trees1, Delay, FraudStatus, State), TotalFees};
                 false -> {Trees1, TotalFees}
             end
     end.
 
-%% TODO: Use height. This is very inefficient for large delays.
-find_prev_key_nodes(Node, N) when N > 0 ->
-    %% Assert
-    key = node_type(Node),
-    find_prev_key_nodes({ok, Node}, N, none).
-
-find_prev_key_nodes(error,_N,_Acc) ->
-    error;
-find_prev_key_nodes({ok, Node}, 0, Acc) ->
-    {Node, Acc};
-find_prev_key_nodes({ok, Node}, 1,_Acc) ->
-    find_prev_key_nodes(db_find_node(prev_key_hash(Node)), 0, Node);
-find_prev_key_nodes({ok, Node}, N, Acc) ->
-    find_prev_key_nodes(db_find_node(prev_key_hash(Node)), N - 1, Acc).
-
-grant_fees(Node, Trees, Delay, Fees, State) ->
-    {KeyNode1, KeyNode2} = find_prev_key_nodes(Node, Delay + 1),
-    KeyFees = case Node =:= KeyNode2 of
-                  true  -> Fees;
-                  false -> db_get_fees(hash(KeyNode2))
-              end,
+find_predecessor_at_height(Node, Height) ->
+    case node_height(Node) of
+        Height -> Node;
+        H when H < Height -> error({cannot_find_predecessor, H, Height});
+        H ->
+            case db_find_key_nodes_at_height(Height) of
+                {ok, [KeyNode]} -> KeyNode;
+                {ok, KeyNodes} -> find_one_predecessor(KeyNodes, Node);
+                error -> error({cannot_find_predecessor, H, Height})
+            end
+    end.
+
+find_one_predecessor([N|Left], Node) ->
+    Hash1 = hash(N),
+    Hash2 = hash(Node),
+    case find_fork_point(Hash1, Hash2) of
+        {ok, Hash1} -> N;
+        {ok, _} -> find_one_predecessor(Left, Node);
+        error -> find_one_predecessor(Left, Node)
+    end.
+
+grant_fees(Node, Trees, Delay, FraudStatus, State) ->
+    KeyNode4 = find_predecessor_at_height(Node, node_height(Node) - Delay + 2),
+    KeyNode3 = db_get_node(prev_key_hash(KeyNode4)),
+    KeyNode2 = db_get_node(prev_key_hash(KeyNode3)),
+    KeyNode1 = db_get_node(prev_key_hash(KeyNode2)),
+    FraudStatus1 = db_get_fraud_status(hash(KeyNode3)),
+    FraudStatus2 = case KeyNode4 =:= Node of
+                       true  -> FraudStatus;
+                       false -> db_get_fraud_status(hash(KeyNode4))
+                   end,
+    KeyFees  = db_get_fees(hash(KeyNode2)),
```

Comment:
I see that this key block needs already to be in the DB so ok that Fees was removed from arguments.